File: src/controller.cpp
#include "controller.hpp"

#include <fstream>

#include "constants.hpp"
#include "invariant.hpp"
#include "logger.hpp"
#include "util.hpp"

using namespace std;

Controller::Controller() : _network(this) {}

Controller &Controller::get() {
    static Controller instance;
    return instance;
}

void Controller::init(const string &networkSpec,
                      const string &invSpec,
                      const string &p4configPath,
                      const string &p4infoPath,
                      const string &outputDir) {
    this->_network.init(networkSpec);
    this->_invs = InvariantsParser()(invSpec);
    this->_p4Paths = {p4configPath, p4infoPath};
    this->_outputDir = outputDir;

    // TODO: Calculate the invariant-wide equivalent packet sets
    // self.ps_to_invs = dict() # PacketSet -> [ Invariants ]
    // self.ps_to_dfas = dict() # PacketSet -> DFA
}

void Controller::connectSwitches() {
    for (const auto &[name, sw] : this->_network.switches()) {
        sw->connect();
    }
}

void Controller::installP4Programs() {
    // For now we assume all P4 switches share the same P4 program/config, so we
    // read the files once and pass in the content for all switches. If the
    // assumption changes, where each switch may have a different
    // program/config, we may simply pass in the file paths.
    string p4config, p4info;

    if (this->_p4Paths.p4configPath.empty()) {
        logger.error("P4 config path empty");
    }

    ifstream ifs(this->_p4Paths.p4configPath);
    p4config =
        string(istreambuf_iterator<char>(ifs), istreambuf_iterator<char>());

    if (!this->_p4Paths.p4infoPath.empty()) {
        ifstream ifs(this->_p4Paths.p4infoPath);
        p4info =
            string(istreambuf_iterator<char>(ifs), istreambuf_iterator<char>());
    }

    for (const auto &[name, sw] : this->_network.switches()) {
        sw->installP4Config(p4config, p4info);
    }
}

size_t Controller::installRIB() {
    size_t numRules = 0;

    for (const auto &[nodeName, prefixMap] : this->_network.rib()) {
        auto it = this->_network.nodes().find(nodeName);

        if (it == this->_network.nodes().end()) {
            logger.warn(nodeName + " doesn't exist. Skipping rules");
            continue;
        }

        numRules += it->second->installRules(prefixMap);
    }

    return numRules;
}

size_t Controller::installEncapDecapRules() {
    size_t numRules = 0;

    for (const auto &[swName, sw] : this->_network.switches()) {
        // Only install at the border switches
        if (sw->hostPorts().empty()) {
            continue;
        }

        // Install rules for entering the network
        {
            P4TableEntry entry;
            entry.tableName("MyIngress.encapsulation");
            entry.addMatchField("hdr.ipv4.protocol",
                                {uint_to_be_str(PROTO_VERIFICATION),
                                 uint_to_be_str(PROTO_VERIFICATION)});
            entry.actionName("NoAction");
            entry.priority(Priority::High);
            sw->installTableEntry(entry);
        }
        {
            P4TableEntry entry;
            entry.tableName("MyIngress.encapsulation");
            entry.actionName("MyIngress.insert_verification_header");
            entry.priority(Priority::Low);
            sw->installTableEntry(entry);
        }

        // Install rules for leaving the network
        for (const auto &hostPort : sw->hostPorts()) {
            P4TableEntry entry;
            entry.tableName("MyEgress.check_leaving");
            entry.addMatchField("std_meta.egress_port",
                                {uint_to_be_str(hostPort)});
            entry.actionName("MyEgress.mark_leaving");
            sw->installTableEntry(entry);
        }

        // Install rules for removing verification headers
        {
            P4TableEntry entry;
            entry.tableName("MyEgress.decapsulation");
            entry.addMatchField("meta.verification.leaving",
                                {uint_to_be_str(uint8_t(1))});
            entry.actionName("MyEgress.remove_verification_header");
            sw->installTableEntry(entry);
        }

        numRules += 3 + sw->hostPorts().size();
    }

    return numRules;
}

size_t Controller::installVerificationRules() {
    size_t numRules = 0;

    if (!this->_invs.empty()) {
        numRules += this->installEncapDecapRules();
    }

    // Compile rules
    unordered_map<shared_ptr<Switch>, unordered_set<P4TableEntry>> entriesMap;

    for (const auto &inv : this->_invs) {
        auto em = inv->getTableEntries(this->_network);
        entriesMap.merge(em);
        for (auto &[sw, entries] : em) {
            entriesMap.at(sw).merge(entries);
        }
    }

    // Install rules
    for (auto &[sw, entries] : entriesMap) {
        for (auto &entry : entries) {
            sw->installTableEntry(entry);
        }

        numRules += entries.size();
    }

    return numRules;
}

static const int sigs[] = {SIGHUP, SIGINT, SIGQUIT, SIGTERM};

void Controller::sig_handler(int sig) {
    switch (sig) {
    case SIGHUP:
    case SIGINT:
    case SIGQUIT:
    case SIGTERM: {
        Controller &controller = Controller::get();
        controller.killAllThreads();
        controller.waitForThreads();
        logger.info("Terminated");
        exit(0);
    }
    }
}

void Controller::startProcessPacketIn() {
    // Process each switch asynchronously
    for (const auto &[_, sw] : this->_network.switches()) {
        sw->startProcessPacketIn();
    }

    // Register signal handler
    struct sigaction action;
    action.sa_handler = Controller::sig_handler;
    sigemptyset(&action.sa_mask);
    for (size_t i = 0; i < sizeof(sigs) / sizeof(int); ++i) {
        sigaddset(&action.sa_mask, sigs[i]);
    }
    for (size_t i = 0; i < sizeof(sigs) / sizeof(int); ++i) {
        sigaction(sigs[i], &action, nullptr);
    }
}

void Controller::waitForThreads() {
    for (const auto &[_, sw] : this->_network.switches()) {
        sw->waitForThreads();
    }
}

void Controller::killAllThreads() {
    for (const auto &[_, sw] : this->_network.switches()) {
        sw->killAllThreads();
    }
}

void Controller::start() {

    // TODO: set_reduced_MTU for end hosts
    // if len(self.invariants) > 0:
    //     for host_dict in self.network['hosts'].values():
    //         for intf in host_dict['intfs'].keys():
    //             # MTU: 1500 - verification header length
    //             cmd = 'ip link set dev {} mtu 1497'.format(intf)
    //             run_mnexc_cmd(host_dict['pid'], cmd)

    logger.info("Connect to network");
    this->connectSwitches();

    logger.info("Install P4 programs");
    this->installP4Programs();

    size_t fwdRules = this->installRIB();
    logger.info("Install forwarding rules... " + to_string(fwdRules));

    size_t verRules = this->installVerificationRules();
    logger.info("Install verification rules... " + to_string(verRules));

    logger.info("===========================================");
    logger.info("Start controller-switch event loops");
    logger.info("===========================================");
    this->startProcessPacketIn();
    this->waitForThreads();
}

File: src/controller.py
#!/usr/bin/env python3

import grpc
import json
import time
import asyncio
from collections import defaultdict
from pypacker.layer12 import ethernet
from pypacker.layer3 import ip, icmp

from .p4runtime_lib.bmv2 import Bmv2SwitchConnection
from .p4runtime_lib.helper import P4InfoHelper
from .p4runtime_lib.error_utils import printGrpcError
from .p4runtime_lib.switch import ShutdownAllSwitchConnections
from .constants import *
from .invariants import *
from .network import run_mnexc_cmd


class Controller:

    def __init__(self, network_json, inv_json, p4info_path, bmv2_json):
        # Load the network
        self.network = {}
        with open(network_json, 'r') as infile:
            self.network = json.load(infile)

        # Load the invariants
        self.invariants = InvariantsParser.parse(inv_json)

        # TODO: Calculate the invariant-wide equivalent packet sets
        self.ps_to_invs = dict() # PacketSet -> [ Invariants ]
        self.ps_to_dfas = dict() # PacketSet -> DFA

        # Switch connections
        self.sw_conns = {}
        self.p4info_helper = P4InfoHelper(p4info_path)
        self.bmv2_json = bmv2_json
        self.futures = []

    def set_reduced_MTU(self):
        if len(self.invariants) > 0:
            for host_dict in self.network['hosts'].values():
                for intf in host_dict['intfs'].keys():
                    # MTU: 1500 - verification header length
                    cmd = 'ip link set dev {} mtu 1497'.format(intf)
                    run_mnexc_cmd(host_dict['pid'], cmd)

    def connect(self):
        for sw in self.network['switches'].values():
            p4rt_logs = 'logs/{}-p4runtime-requests.txt'.format(sw['name'])
            sw_conn = Bmv2SwitchConnection(name=sw['name'],
                                           address='127.0.0.1:{}'.format(
                                               sw['grpc_port']),
                                           device_id=sw['device_id'],
                                           proto_dump_file=p4rt_logs)
            self.sw_conns[sw['name']] = sw_conn

            # Send master arbitration update message to establish this
            # controller as master (required by P4Runtime before performing any
            # other write operation)
            sw_conn.MasterArbitrationUpdate()

    def install_P4_programs(self):
        for sw_conn in self.sw_conns.values():
            print('Installing P4 program on', sw_conn.name)
            sw_conn.SetForwardingPipelineConfig(
                p4info=self.p4info_helper.p4info,
                bmv2_json_file_path=self.bmv2_json)

    def install_rules(self, rules_dict):
        num_rules = 0
        for sw_name, rules in rules_dict.items():
            for rule in rules:
                tbl_name = rule['table_name']
                mfs = rule['match_fields'] if 'match_fields' in rule else None
                an = rule['action_name'] if 'action_name' in rule else None
                aps = rule['action_params'] if 'action_params' in rule else None
                priority = rule['priority'] if 'priority' in rule else None
                table_entry = self.p4info_helper.buildTableEntry(
                    table_name=tbl_name,
                    match_fields=mfs,
                    action_name=an,
                    action_params=aps,
                    priority=priority)
                # print(sw_name, json.dumps(rule, indent=4))
                self.sw_conns[sw_name].WriteTableEntry(table_entry)
            num_rules += len(rules)
        return num_rules

    def _install_encap_decap_rules(self):
        num_rules = 0
        for sw_name, sw in self.network['switches'].items():
            # Only install at the border switches
            if len(sw['host_ports']) == 0:
                continue

            # Install rules for entering the network
            table_entry = self.p4info_helper.buildTableEntry(
                table_name='MyIngress.encapsulation',
                match_fields={
                    'hdr.ipv4.protocol':
                        (PROTO_VERIFICATION, PROTO_VERIFICATION)
                },
                action_name='NoAction',
                priority=Priority.HIGH)
            self.sw_conns[sw_name].WriteTableEntry(table_entry)

            table_entry = self.p4info_helper.buildTableEntry(
                table_name='MyIngress.encapsulation',
                match_fields={},
                action_name='MyIngress.insert_verification_header',
                priority=Priority.LOW)
            self.sw_conns[sw_name].WriteTableEntry(table_entry)

            # Install rules for leaving the network
            for host_port in sw['host_ports']:
                table_entry = self.p4info_helper.buildTableEntry(
                    table_name='MyEgress.check_leaving',
                    match_fields={
                        'std_meta.egress_port': host_port,
                    },
                    action_name='MyEgress.mark_leaving',
                )
                self.sw_conns[sw_name].WriteTableEntry(table_entry)

            # Install rules for removing verification headers
            table_entry = self.p4info_helper.buildTableEntry(
                table_name='MyEgress.decapsulation',
                match_fields={'meta.verification.leaving': 1},
                action_name='MyEgress.remove_verification_header')
            self.sw_conns[sw_name].WriteTableEntry(table_entry)

            num_rules += 3 + len(sw['host_ports'])
        return num_rules

    # def _install_trace_rules(self):
    #     for sw_name, sw in self.network['switches'].items():
    #         table_entry = self.p4info_helper.buildTableEntry(
    #             table_name='MyIngress.trace',
    #             match_fields={
    #                 'hdr.verification.traceCount': (0, TRACE_LENGTH - 1),
    #                 'meta.verification.leaving': 0
    #             },
    #             action_name='MyIngress.add_trace',
    #             action_params={
    #                 'add': 1,
    #                 'swId': sw['device_id']
    #             },
    #             priority=Priority.LOW)
    #         self.sw_conns[sw_name].WriteTableEntry(table_entry)

    #         table_entry = self.p4info_helper.buildTableEntry(
    #             table_name='MyIngress.trace',
    #             match_fields={
    #                 'hdr.verification.traceCount': (TRACE_LENGTH, TRACE_LENGTH),
    #                 'meta.verification.leaving': 0
    #             },
    #             action_name='MyIngress.add_trace',
    #             action_params={
    #                 'add': 0,
    #                 'swId': sw['device_id']
    #             },
    #             priority=Priority.LOW)
    #         self.sw_conns[sw_name].WriteTableEntry(table_entry)

    def install_verification_rules(self):
        num_rules = 0

        if len(self.invariants) > 0:
            num_rules += self._install_encap_decap_rules()
            # num_rules += self._install_trace_rules() # Disable for now

        installed_inv_rules = defaultdict(set)

        def _is_installed(sw_name, rule):
            nonlocal installed_inv_rules
            match_entry = {k: rule[k] for k in ['table_name', 'match_fields']}
            return (json.dumps(match_entry, sort_keys=True)
                    in installed_inv_rules[sw_name])

        def _add_installed_rule(sw_name, rule):
            nonlocal installed_inv_rules
            match_entry = {k: rule[k] for k in ['table_name', 'match_fields']}
            installed_inv_rules[sw_name].add(
                json.dumps(match_entry, sort_keys=True))

        compile_time = 0
        install_time = 0

        for invariant in self.invariants:
            print('Installing rules for invariant', invariant.name)
            # Compile the rules
            start = time.perf_counter()
            rules = invariant.get_rules(self.network)
            end = time.perf_counter()
            compile_time += end - start
            # Filter out duplicate rules
            rules = {
                sw_name: [
                    rule for rule in rules_list
                    if not _is_installed(sw_name, rule)
                ] for sw_name, rules_list in rules.items()
            }
            # Install rules
            start = time.perf_counter()
            num_rules += self.install_rules(rules)
            end = time.perf_counter()
            install_time += end - start
            # Remember the installed rules
            for sw_name, rules_list in rules.items():
                for rule in rules_list:
                    _add_installed_rule(sw_name, rule)

        print('Verification rules:', num_rules)
        print('Compile time:', compile_time, 'seconds')
        print('Install time:', install_time, 'seconds')

    def read_table_rules(self, switch_name):
        sw_conn = self.sw_conns[switch_name]
        print('\n----- Reading tables rules for %s -----' % sw_conn.name)
        for response in sw_conn.ReadTableEntries():
            for entity in response.entities:
                entry = entity.table_entry
                print('-----')
                table_name = self.p4info_helper.get_tables_name(entry.table_id)
                print('%s: ' % table_name, end=' ')
                for m in entry.match:
                    print(self.p4info_helper.get_match_field_name(
                        table_name, m.field_id),
                          end=' ')
                    print('%r' % (self.p4info_helper.get_match_field_value(m),),
                          end=' ')
                action = entry.action.action
                action_name = self.p4info_helper.get_actions_name(
                    action.action_id)
                print('->', action_name, end=' ')
                for p in action.params:
                    print(self.p4info_helper.get_action_param_name(
                        action_name, p.param_id),
                          end=' ')
                    print('%r' % p.value, end=' ')
                print()

    def process_packet_in(self, sw_conn, packet):
        invId = int.from_bytes(packet.metadata[0].value, byteorder='big')
        pkt_bytes = packet.payload
        eth = ethernet.Ethernet(pkt_bytes)
        print('Violation occurred at', sw_conn.name)
        print('Violated invariant:')
        print(self.invariants[invId])
        print(eth)

    def process_switch(self, sw_conn):
        try:
            while True:
                result = sw_conn.PacketIn()
                if result.WhichOneof('update') == 'packet':
                    self.process_packet_in(sw_conn, result.packet)
        except grpc.RpcError as e:
            printGrpcError(e)
        except Exception as e:
            print(e)

    async def async_start(self):
        try:
            self.set_reduced_MTU()
            self.connect()
            self.install_P4_programs()
            print('Installing forwarding rules')
            self.install_rules(self.network['rules'])
            print('Installing verification rules')
            self.install_verification_rules()
        except grpc.RpcError as e:
            printGrpcError(e)
            raise
        except Exception as e:
            print(e)
            raise

        print('===========================================')
        print('Start controller-switch event loops')
        print('===========================================')

        # Process each switch asynchronously
        loop = asyncio.get_running_loop()
        for sw_conn in self.sw_conns.values():
            self.futures.append(
                loop.run_in_executor(None, self.process_switch, sw_conn))

    def start(self):
        try:
            asyncio.run(self.async_start())
        except:
            print('Shutting down')
            self.shutdown()

    def shutdown(self):
        try:
            for future in self.futures:
                future.cancel()
            # ShutdownAllSwitchConnections()
        except:
            pass
